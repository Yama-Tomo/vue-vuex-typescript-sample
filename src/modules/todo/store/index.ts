import type { Module } from 'vuex';
import { RootState } from '../../module_mapper';
import defaultState, { initialStateResolver, State } from './state';
import mutations from './mutations';
import actions, { ActionTree as TodoActTree } from './actions';
import getters, { GetterTree as TodoGetterTree } from './getters';

type ModuleNameSpace = 'todoModule';

export * from './state';

export const module = (
  initialState?: unknown
): Record<ModuleNameSpace, Module<State, RootState>> => {
  const state = initialState
    ? initialStateResolver(initialState)
    : defaultState();

  return {
    todoModule: {
      namespaced: true,
      state,
      mutations,
      actions,
      getters,
    },
  };
};

export type StateTree = Record<ModuleNameSpace, State>;
export type GetterTree = Record<ModuleNameSpace, TodoGetterTree>;
export type ActionTree = Record<ModuleNameSpace, TodoActTree>;
