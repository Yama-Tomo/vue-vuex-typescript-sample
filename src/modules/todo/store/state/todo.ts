export default class Todo {
  public text: string;
  public done: boolean;

  constructor(text: string, done: boolean) {
    this.text = text;
    this.done = done;
  }
}
