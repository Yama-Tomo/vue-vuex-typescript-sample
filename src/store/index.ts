export * from './helper';
export * from './module_mapper';
