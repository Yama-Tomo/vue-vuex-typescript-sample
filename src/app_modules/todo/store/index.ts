import Todo from './state/todo';
import { state, TodoState } from './state';
import { mutations } from './mutations';
import { actions } from './actions';
import { getters } from './getters';

export const todoStore = {
  namespaced: true,
  state,
  mutations,
  actions,
  getters,
};

export const initialStateResolver = (values: any): TodoState => {
  const todos: Todo[] = (values.todos as any[] || []).filter((value) => {
    return value.hasOwnProperty('text') && value.hasOwnProperty('done');
  }).map((value) => {
    return new Todo(value.text.toString(), !!value.done);
  });

  return { todos };
};
