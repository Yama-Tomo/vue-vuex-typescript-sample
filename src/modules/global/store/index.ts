import { state } from './state';
import { mutations } from './mutations';
import { actions } from './actions';
import { getters } from './getters';

export const globalStore = {
  namespaced: true,
  state,
  mutations,
  actions,
  getters,
};
