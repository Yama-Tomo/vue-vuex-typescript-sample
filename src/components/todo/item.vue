<script lang="tsx">
import Vue, { PropType, VNode } from 'vue';
import * as vts from 'vue-tsx-support';
import { mdiDelete } from '@mdi/js';
import { ActionTree, StateTree } from '@/store';
import { InputEvent, ComponentProps } from '@/types';

// eslint-disable-next-line no-use-before-define
const liClass = (todo: Props['todo'], editing: boolean) => {
  const stack = ['todo'];

  if (todo.done) {
    stack.push('completed');
  }

  if (editing) {
    stack.push('editing');
  }

  return stack.join(' ');
};

type LocalState = {
  editing: boolean;
};

const Component = Vue.extend({
  props: {
    todo: {
      type: Object as PropType<StateTree['todo']['todos'][number]>,
      default: undefined,
    },
    actions: {
      type: Object as PropType<ActionTree['todo']>,
      default: undefined,
    },
  },
  data(): LocalState {
    return {
      editing: false,
    };
  },
  methods: {
    changeEditMode() {
      this.editing = true;
      this.$nextTick(() => (this.$refs.inputText as HTMLInputElement).focus());
    },
    editTodo(text: string) {
      return this.actions.editTodo({ todo: this.todo, text });
    },
    toggleTodo() {
      this.actions.toggleTodo({ todo: this.todo });
    },
    removeTodo() {
      this.actions.removeTodo(this.todo);
    },
    onKeyPress(e: Event) {
      const keyCode = (e as KeyboardEvent).key;
      if (keyCode === 'Enter') {
        this.doneEdit(e);
      }
    },
    onKeyUp(e: Event) {
      const keyCode = (e as KeyboardEvent).key;
      if (keyCode === 'Escape') {
        this.cancelEdit(e);
      }
    },
    doneEdit(e: Event) {
      const value = (e as InputEvent).target.value.trim();

      if (!value) {
        this.removeTodo();
      } else if (this.editing) {
        this.editTodo(value);
        this.editing = false;
      }
    },
    cancelEdit(e: Event) {
      (e as InputEvent).target.value = this.todo.text;
      this.editing = false;
    },
  },
  render(): VNode {
    return (
      <v-list-item class={liClass(this.todo, this.editing)}>
        <v-list-item-content>
          <v-checkbox
            class="shrink mr-2 mt-0"
            hide-details
            input-value={this.todo.done}
            onChange={() => this.toggleTodo()}
          />
          <div
            class="text"
            onClick={() => this.changeEditMode()}
            v-show={!this.editing}
          >
            {this.todo.text}
          </div>
          <v-text-field
            ref="inputText"
            value={this.todo.text}
            v-show={this.editing}
            onKeyup={(e: Event) => this.onKeyUp(e)}
            onKeypress={(e: Event) => this.onKeyPress(e)}
            onBlur={(e: Event) => this.doneEdit(e)}
            hide-details
          />
        </v-list-item-content>
        <v-list-item-action>
          <v-btn icon onClick={() => this.removeTodo()}>
            <v-icon>{mdiDelete}</v-icon>
          </v-btn>
        </v-list-item-action>
      </v-list-item>
    );
  },
});

type Props = ComponentProps<typeof Component>;
export default vts.ofType<Props>().convert(Component);
</script>

<style lang="scss" scoped>
.todo {
  &.editing {
    background-color: #fafafa;
  }

  &.completed {
    .text {
      text-decoration: line-through;
    }
  }

  .v-list-item__content {
    overflow: inherit;

    .v-input {
      padding-top: 0;
      margin-top: 0;
    }

    .text {
      display: flex;
      flex: 1;
      cursor: pointer;
    }
  }
}
</style>
