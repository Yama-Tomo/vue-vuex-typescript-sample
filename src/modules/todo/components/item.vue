<script lang="tsx">
import { Component, Vue, Prop } from 'vue-property-decorator';
import * as vts from 'vue-tsx-support';
import { HTMLElementEvent } from '../../../types/dom';
import { Todo } from '@/modules/todo/store';
import { ActionTree } from '@/modules/module_mapper';

@Component
export class Item extends Vue {
  @Prop()
  public todo!: Todo;

  @Prop()
  public actions!: ActionTree['todoModule'];

  public editing = false;

  public render() {
    return (
      <li
        class={[
          'todo',
          ...(this.todo.done ? ['completed'] : []),
          ...(this.editing ? ['editing'] : []),
        ].join(' ')}
      >
        <div class="view">
          <input
            class="toggle"
            type="checkbox"
            checked={this.todo.done}
            onChange={() => this.toggleTodo()}
          />
          <label onDblclick={() => this.changeEditMode()}>
            {this.todo.text}
          </label>
          <button class="destroy" onClick={() => this.removeTodo()}>
            remove
          </button>
        </div>
        <input
          class="edit"
          ref="inputText"
          value={this.todo.text}
          v-show={this.editing}
          onKeyup={e => this.onKeyup(e)}
          onBlur={e => this.doneEdit(e)}
        />
      </li>
    );
  }

  public changeEditMode() {
    this.editing = true;
    this.$nextTick(() => (this.$refs.inputText as HTMLInputElement).focus());
  }

  public editTodo(text: string) {
    return this.actions.editTodo({ todo: this.todo, text });
  }

  public toggleTodo() {
    this.actions.toggleTodo(this.todo);
  }

  public removeTodo() {
    this.actions.removeTodo(this.todo);
  }

  public onKeyup(e: Event) {
    const keyCode = (e as KeyboardEvent).key;
    if (keyCode === 'Enter') {
      this.doneEdit(e);
    }

    if (keyCode === 'Escape') {
      this.cancelEdit(e);
    }
  }

  public doneEdit(e: Event) {
    const value = (e as HTMLElementEvent<HTMLInputElement>).target.value.trim();

    if (!value) {
      this.removeTodo();
    } else if (this.editing) {
      this.editTodo(value);
      this.editing = false;
    }
  }

  public cancelEdit(e: Event) {
    (e as HTMLElementEvent<HTMLInputElement>).target.value = this.todo.text;
    this.editing = false;
  }
}

export default vts.ofType<Partial<Item>>().convert(Item);
</script>

<style lang="scss" scoped>
label {
  padding-right: 10px;
  padding-left: 10px;
}

.editing {
  label {
    color: #888;
  }
}

.completed {
  label {
    text-decoration: line-through;
  }
}
</style>
