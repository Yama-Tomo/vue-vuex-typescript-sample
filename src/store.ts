import Vue from 'vue';
import Vuex from 'vuex';
import { modules } from './namespace_maps';
import { todoStore } from './modules/todo/store';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    [modules.todo]: todoStore,
  },
});
