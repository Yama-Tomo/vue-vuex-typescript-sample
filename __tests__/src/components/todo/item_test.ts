/**
 * @jest-environment jsdom
 */

import Vue from 'vue';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import * as Component from '@/components/todo/item.vue';
import { actions, Todo } from '@/store/todo';
Vue.use(Vuetify);

const localVue = createLocalVue();

const defaultTodoState = { text: 'aaa', done: false };
const getWrapper = (arg?: {
  actions?: { [KEY in keyof typeof actions]?: any };
  state?: Todo;
  data?: () => any;
}) => {
  const options = {
    data: arg && arg.data ? arg.data : () => ({}),
    localVue,
    vuetify: new Vuetify(),
    propsData: {
      todo: {
        ...defaultTodoState,
        ...(arg && arg.state ? arg.state : {}),
      },
      actions: arg && arg.actions,
    },
    mocks: {
      $t: (key: string) => `[Mocked i18n] ${key}`,
    },
  };

  return mount(Component.default, options);
};

describe('behaviors', () => {
  test('チェックボックスのチェックを入れたらステートを完了状態に変更するvuexのactionを呼び出すこと', () => {
    const actionMock = jest.fn();
    const wrapper = getWrapper({ actions: { toggleTodo: actionMock } });

    wrapper.find('input[type="checkbox"]').trigger('click');
    expect(actionMock.mock.calls[0]).toEqual([{ todo: defaultTodoState }]);
  });

  test('ラベルをクリックしたら編集モードになること', async () => {
    const wrapper = getWrapper();

    wrapper.find('.text').trigger('click');

    await Vue.nextTick();
    expect(wrapper.find('.v-text-field input').isVisible()).toBe(true);
    expect(wrapper.find('.v-list-item').classes()).toContain('editing');
  });

  test('削除ボタンを押された時にステートを削除するvuexのactionを呼び出すこと', () => {
    const actionMock = jest.fn();
    const wrapper = getWrapper({ actions: { removeTodo: actionMock } });

    wrapper.find('button').trigger('click');
    expect(actionMock.mock.calls[0]).toEqual([defaultTodoState]);
  });

  test('todoの入力があった時にステートを変更するvuexのactionを呼び出すこと', () => {
    const actionMock = jest.fn();
    const wrapper = getWrapper({
      actions: { editTodo: actionMock },
      data() {
        return { editing: true };
      },
    });

    const appendText = '--append';
    const changeVal = defaultTodoState.text + appendText;
    const inputEl = wrapper.find('.v-text-field input');

    inputEl.trigger('keyup', { key: appendText });
    inputEl.setValue((inputEl.element as HTMLInputElement).value + appendText);
    inputEl.trigger('keypress', { key: 'Enter' });
    expect(actionMock.mock.calls[0]).toEqual([
      { todo: defaultTodoState, text: changeVal },
    ]);
  });

  test('todoの入力があった時にescを押されたらテキストボックスが非表示になること', async () => {
    const wrapper = getWrapper({
      data() {
        return { editing: true };
      },
    });

    const inputEl = wrapper.find('.v-text-field input');
    inputEl.trigger('keyup', { key: 'Escape' });
    await Vue.nextTick();

    expect(inputEl.isVisible()).toBe(false);
  });

  test('todoを未入力状態にしてenterが押されたらステートを削除するvuexのactionを呼び出すこと', () => {
    const actionMock = jest.fn();
    const wrapper = getWrapper({
      data() {
        return { editing: true };
      },
      actions: { removeTodo: actionMock },
    });

    const inputEl = wrapper.find('.v-text-field input');
    inputEl.setValue('');
    inputEl.trigger('keypress', { key: 'Enter' });
    expect(actionMock.mock.calls[0]).toEqual([defaultTodoState]);
  });
});

describe('unit-tests', () => {
  test('liのclass属性のテスト', () => {
    const func = (Component as any).__get__('liClass');
    expect(func({ text: 'aaa', done: true }, true)).toEqual(
      'todo completed editing'
    );
    expect(func({ text: 'aaa', done: true }, false)).toEqual('todo completed');
  });
});

describe('rendering', () => {
  test('完了しているTODOは完了を表すcssが付いていること', () => {
    const wrapper = getWrapper({
      state: { text: 'hogehoge', done: true },
    });

    expect(wrapper.find('.v-list-item').classes()).toContain('completed');
  });

  test('renders correctly', () => {
    const wrapper = getWrapper({
      data() {
        return { editing: true };
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
