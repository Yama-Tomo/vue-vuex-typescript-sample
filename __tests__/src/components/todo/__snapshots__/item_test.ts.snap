// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<div
  class="v-list-item theme--light todo editing"
  tabindex="-1"
>
  <div
    class="v-list-item__content"
  >
    <div
      class="v-input v-input--hide-details theme--light v-input--selection-controls v-input--checkbox shrink mr-2 mt-0"
    >
      <div
        class="v-input__control"
      >
        <div
          class="v-input__slot"
        >
          <div
            class="v-input--selection-controls__input"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"
            />
            <input
              aria-checked="false"
              id="input-59"
              role="checkbox"
              type="checkbox"
              value=""
            />
            <div
              class="v-input--selection-controls__ripple"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="text"
      style="display: none;"
    >
      aaa
    </div>
    <div
      class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field"
    >
      <div
        class="v-input__control"
      >
        <div
          class="v-input__slot"
        >
          <div
            class="v-text-field__slot"
          >
            <input
              id="input-61"
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="v-list-item__action"
  >
    <button
      class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <span
          aria-hidden="true"
          class="v-icon notranslate theme--light"
        >
          <svg
            aria-hidden="true"
            class="v-icon__svg"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19,4H15.5L14.5,3H9.5L8.5,4H5V6H19M6,19A2,2 0 0,0 8,21H16A2,2 0 0,0 18,19V7H6V19Z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;
